### PR TITLE
MicrodiffZoom handles B-Zoom and MotorisedZoom.

### PR DESCRIPTION
**Pull request by beteva**
Mockup handles specific values format.

### CODE REVIEW

File: HardwareObjects/mockup/MicrodiffZoomMockup.py [hunk 1]
```diff
@@ -1,38 +1,64 @@
...
     def init(self):
         """Initialize the zoom"""
         AbstractNState.init(self)
 
-        try:
-            values = ast.literal_eval(self.getProperty("values")).values()
-        except AttributeError:
-            limits = (0, 10)
+        self.initialise_values()
+        _len = self.VALUES.__members__.__len__() - 1
```
Comment (rhfogh) [inline HardwareObjects/mockup/MicrodiffZoomMockup.py:29]: _len = len(self.VALUES) - 1
should give the same result

File: HardwareObjects/MicrodiffZoom.py [hunk 2]
```diff
@@ -40,51 +42,66 @@
...
-            if _high == float("inf"):
-                _high = 10
-
-            self.set_limits((_low, _high))
-        except ValueError:
-            self.set_limits((1, 10))
-
         self.initialise_values()
+        # check if we have values other that UKNOWN
+        _len = self.VALUES.__members__.__len__() - 1
```
Comment (rhfogh) [inline HardwareObjects/MicrodiffZoom.py:55]: _len = len(self.VALUES) - 1
should give the same result

**Comment by rhfogh** [comment]
Looks good.

The len(self.VALUES) stuff is about clarity, not funcionality.

One question: It looks from the code that it is assumed that the limits are integers (or maybe None). Should this not be reflected in the function documentation, so it mentioned tuples of 'int' instead of tuples of 'elements'.

If this is so, it might reflect also on the documentation of AbstractActuator - if limits are either ints of floats or None (and can they be anything else), maybe htat should be rflected in the documentation also there.

One idea:
We could use IntEnum for AbstractZoom.VALUE, or maybe even for AbstractNState.VALUE in general. An intEnum works arithmetically a an integer, so the limits could be the actual enumeration members, and you could do comparison, tests and sorting on them as if they were inegers.

File: HardwareObjects/mockup/MicrodiffZoomMockup.py (hunk 1 above)
Comment (beteva) [reply HardwareObjects/mockup/MicrodiffZoomMockup.py:29]: OK

File: HardwareObjects/MicrodiffZoom.py (hunk 2 above)
Comment (beteva) [reply HardwareObjects/MicrodiffZoom.py:55]: OK*2

**Comment by beteva** [comment]
> One question: It looks from the code that it is assumed that the limits are integers (or maybe None). Should this not be reflected in the function documentation, so it mentioned tuples of 'int' instead of tuples of 'elements'.

You are right. i did change, but apparently not everywhere. 
> If this is so, it might reflect also on the documentation of AbstractActuator - if limits are either ints of floats or None (and can they be anything else), maybe htat should be rflected in the documentation also there.
> 
This is specific fro the MD2 Zoom. I would only document it in the MicrodifZoom.py
> One idea:
> We could use IntEnum for AbstractZoom.VALUE, or maybe even for AbstractNState.VALUE in general. An intEnum works arithmetically a an integer, so the limits could be the actual enumeration members, and you could do comparison, tests and sorting on them as if they were inegers.

Unfortunately not, as we have the "UNKNOWN" value, which is a string and I would say , better stay this way. Otherwise we have to think about what would be a value, which represents the UNKNOWN, but this is tricky.

**Comment by rhfogh** [comment]
@beteva 
Makes sense.

**Comment by marcus-oscarsson** [comment]
Wonderful ! :)

**Pull request merged**